[CI] ShuffleForcedMergePolicyTests testDiagnostics failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #6102 / rocky-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/35atahghgt7ce)
- [elasticsearch-pull-request #54240 / part-1](https://gradle-enterprise.elastic.co/s/p6nixutlnepmu)
- [elasticsearch-periodic-platform-support #6081 / ubuntu-2004_platform-support-unix](https://gradle-enterprise.elastic.co/s/apcmh5zzj3i3e)

**Reproduction Line:**
```
null
```

**Applicable branches:**
9.0

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.index.engine.ShuffleForcedMergePolicyTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testDiagnostics),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: null
```

**Issue Reasons:**
- [9.0] 3 failures in test testDiagnostics (1.9% fail rate in 156 executions)
- [9.0] 2 failures in pipeline elasticsearch-periodic-platform-support (50.0% fail rate in 4 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -291,9 +291,6 @@ tests:
 - class: org.elasticsearch.env.NodeEnvironmentTests
   method: testGetBestDowngradeVersion
   issue: https://github.com/elastic/elasticsearch/issues/121316
-- class: org.elasticsearch.index.engine.ShuffleForcedMergePolicyTests
-  method: testDiagnostics
-  issue: https://github.com/elastic/elasticsearch/issues/121336
 - class: org.elasticsearch.smoketest.DocsClientYamlTestSuiteIT
   method: test {yaml=reference/rest-api/security/invalidate-tokens/line_194}
   issue: https://github.com/elastic/elasticsearch/issues/121337

--- a/server/src/test/java/org/elasticsearch/index/engine/ShuffleForcedMergePolicyTests.java
+++ b/server/src/test/java/org/elasticsearch/index/engine/ShuffleForcedMergePolicyTests.java
@@ -20,6 +20,7 @@ import org.apache.lucene.index.LeafReader;
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.MergePolicy;
 import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.index.TieredMergePolicy;
 import org.apache.lucene.search.Sort;
 import org.apache.lucene.search.SortField;
 import org.apache.lucene.store.Directory;
@@ -37,7 +38,14 @@ public class ShuffleForcedMergePolicyTests extends BaseMergePolicyTestCase {
             IndexWriterConfig iwc = newIndexWriterConfig();
             // Disable merging on flush.
             iwc.setMaxFullFlushMergeWaitMillis(0L);
-            MergePolicy mp = new ShuffleForcedMergePolicy(newTieredMergePolicy());
+            // Even though we set setMaxFullFlushMergeWaitMillis=0, opening the DirectoryReader
+            // might trigger a merge after flushing the in-memory documents. Therefore, we set
+            // a high enough number of maxSegmentsPerTier (we index at most 300 documents, and we flush
+            // a new segment per 10 documents) that would prevent merging all the segments into one and
+            // making the force merge a no-op.
+            var tieredMergePolicy = new TieredMergePolicy();
+            tieredMergePolicy.setSegmentsPerTier(100);
+            MergePolicy mp = new ShuffleForcedMergePolicy(tieredMergePolicy);
             iwc.setMergePolicy(mp);
             boolean sorted = random().nextBoolean();
             if (sorted) {

